gateway-messages/src/sp_impl.rs

```diff
@@ -209,10 +251,84 @@ pub fn handle_message<H: SpHandler>(
 
     // We know `response` is well-formed and fits into `out` (since it's
     // statically sized for `SpMessage`), so we can unwrap serialization.
-    match hubpack::serialize(&mut out[..], &response) {
+    let mut n = match hubpack::serialize(&mut out[..], &response) {
         Ok(n) => n,
         Err(_) => panic!(),
+    };
+
+    // Append any outgoing trailing data.
+    n += match outgoing_trailing_data {
+        Some(OutgoingTrailingData::DeviceInventory {
+            device_index,
+            total_devices,
+        }) => encode_device_inventory(
+            &mut out[n..],
+            device_index,
+            total_devices,
+            handler,
+        ),
+        None => 0,
+    };
+
+    n
+}
+
+/// Pack as many device description TLV triples as we can into `out`, starting
+/// at `device_index`.
+fn encode_device_inventory<H: SpHandler>(
+    mut out: &mut [u8],
+    mut device_index: u32,
+    total_devices: u32,
+    handler: &mut H,
+) -> usize {
+    use crate::DeviceDescription as DeviceDescriptionHeader;
+
+    let mut total_tlv_len = 0;
+    while device_index < total_devices {
+        let dev = handler.device_description(device_index);
+
+        // Will the serialized description of this device fit in `out`?
+        let len = tlv::tlv_len(
+            DeviceDescriptionHeader::MAX_SIZE
+                + dev.device.len()
+                + dev.description.len(),
+        );
+        if len > out.len() {
+            break;
+        }
+
+        // It will fit: serialize and encode it.
+        match tlv::encode::<_, Infallible>(
+            out,
+            DeviceDescriptionHeader::TAG,
+            |buf| {
+                let header = DeviceDescriptionHeader::from(dev);
+                // We know our buffer is large enough from our length check
+                // above, so this serialization can't fail.
+                let mut n = hubpack::serialize(buf, &header).unwrap();
+
+                // Pack in the device and description
+                for s in [dev.device, dev.description] {
+                    buf[n..][..s.len()].copy_from_slice(s.as_bytes());
+                    n += s.len();
+                }
+
+                Ok(n)
+            },
+        ) {
+            Ok(n) => {
+                total_tlv_len += n;
+                out = &mut out[n..];
+            }
+            // We checked the length above; this error isn't possible.
+            Err(tlv::EncodeError::BufferTooSmall) => panic!(),
+            Err(tlv::EncodeError::Custom(infallible)) => match infallible {},
```

Comment:
I don't really understand what's going on here. Can you enlighten me just for learning purposes? Thanks!

Reply:
Yeah, sorry, this is tricky enough I should put a comment here. `infallible` here has the type [`core::convert::Infallible`](https://doc.rust-lang.org/core/convert/enum.Infallible.html), "the error type for errors that can never happen". It's an uninhabitable enum, which means an instance of it can never exist. Here we appear to have one (in `infallible`), but we can attempt to `match` on it (with no arms, because it's an uninhabitable enum!). The compiler realizes that the match with no arms means `infallible` can't actually exist, which means from type system point of view it can be any type at all:

```rust
// This is fine.
fn infallible_to_u32(infallible: core::convert::Infallible) -> u32 {
    match infallible {}
}

// This is also fine.
fn infallible_to_string(infallible: core::convert::Infallible) -> String {
    match infallible {}
}

// Even this is fine.
fn infallible_to_anything<T>(infallible: core::convert::Infallible) -> T {
    match infallible {}
}
```

This is why it's really useful as an error type; you can unwrap without `.unwrap()`, and the compiler knows the error branch isn't possible:

```rust
fn unwrap_without_panic<T>(result: Result<T, core::convert::Infallible>) -> T {
    match result {
        Ok(t) => t,
        Err(infallible) => match infallible {},
    }
}
```

The actual code here is essentially doing this, but slightly more convoluted because it's only applicable for `EncodeError::Custom(_)`.

Reply:
Thanks for the explanation. I guess you could also use `()` as the error type and always unwrap, but that would generate more code and the branch could make it slower. So this is an optimization, right?

Reply:
Right, this entire branch should not exist in the binary.